### PR TITLE
110 add model causalformer

### LINKED ISSUE #110: Add Model: CausalFormer
### Model Description

This is a Temporal Fusion Transformer with adaptations to discover causally related time-series. The paper from Kong et al. 2024 describe the details. 

## References
Kong, Lingbai, Wengen Li, Hanchen Yang, Yichao Zhang, Jihong Guan, and Shuigeng Zhou. 2024. “CausalFormer: An Interpretable Transformer for Temporal Causal Discovery.” arXiv [Cs.LG]. arXiv. http://arxiv.org/abs/2406.16708.

### Tasks

- [ ] Model saving/loading script
- [ ] Model inference script (API or example)
- [ ] Model performance evaluation script
- [ ] Tutorial demonstrating usage

### Additional Considerations

This relates closely to #108 #107 

### license

yes

### Submission Checklist

- [ ] I have read the contribution guidelines.
- [ ] I have provided a detailed model description.
- [ ] All required scripts are provided, functional, and tested.
- [ ] I have clarified dependencies and hardware considerations.
- [ ] The tutorial is clear and demonstrates model usage.

### PR DESCRIPTION
**Pull request by boersmamarcel**
**Dataset Addition Merge Checklist**

**Please ensure the following are completed before merging this pull request:**

- [x] Contributor signed the license agreement: To contribute to our repository, the contributor must sign the license agreement
- [ ] Thorough Description: The contribution is adequately described, including origin, content, size, licensing, and the relevant academic paper.
- [ ] Clear notebook guide on how to use the script.

Thanks for contributing

## Summary by CodeRabbit

- **New Features**
  - Added explainable AI layers including Dropout, LayerNorm, LeakyReLU, ReLU, Softmax, Linear, and einsum with relevance propagation support.
  - Introduced ValueCaptureHook for capturing intermediate tensors during forward passes to enable explainability.
  - Added causal explanation class for gradient-based relevance propagation in causal models.
  - Enhanced TemporalFusionTransformer and its module with improved input validation and handling of optional networks, plus flexible multi-series data handling.
  - Added comprehensive CausalFormer model suite with multi-head attention, causal convolution, embedding, encoder layers, and explainability support.
  - Introduced new dataset classes and data modules with standardized dictionary-based data formats and custom collate functions for synthetic, AirPassenger, Electricity, StockTicker, and causal datasets.
  - Added modular example scripts and tutorials for causal data exploration, hyperparameter tuning, synthetic data generation, and time series forecasting workflows.
  - Added CausalFormerModule Lightning integration supporting multi-series handling and flexible series selection/aggregation.
  - Added smoke test framework for automated validation of example scripts execution.
  - Added extensive test suites covering explainable layers, models, datasets, integration, performance, and stress testing.
  - Improved robustness and fallback mechanisms in data downloading and dataset loading, including synthetic data generation fallback for stock data.
  - Updated documentation with detailed descriptions of datasets, base classes, tutorials, and testing guidelines.

- **Bug Fixes**
  - Fixed error handling in callback validation for Ray Tune integration.
  - Improved handling of empty inputs and missing features in VariableSelectionNetwork and other modules.
  - Enhanced dataset removal and download error handling with graceful fallbacks and informative messages.

- **Documentation**
  - Added new documentation files for base datasets, synthetic datasets, and comprehensive module test coverage.
  - Updated tutorial notebooks covering TFT, causal models, and time series datasets with improved formatting and clarity.
  - Provided README and usage examples for causal data modules and example scripts.

- **Tests**
  - Added unit tests for explainable AI layers, including einsum relevance propagation.
  - Expanded model tests with fixtures, parameterized tests, and coverage for training, validation, prediction, and edge cases.
  - Added dataset format validation tests ensuring compliance with standardized data interfaces.
  - Included integration and performance tests for core time series forecasting modules.
  - Added tests covering custom collate functions and multi-series data handling robustness.

- **Chores**
  - Updated dependencies and development requirements with new packages and version upgrades.
  - Modified Makefile targets for testing and requirements compilation.
  - Added .gitignore entry for PyTorch Lightning logs.
  - Applied code style and formatting improvements across notebooks and scripts.

### CODE REVIEW

File: fintorch/models/timeseries/causalformer/CausalConvolution.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+                                for i in range(length_input_window):
+                                    # Apply the learned kernel onto the input value x
+                                    # sum over the input window length such that we obtain a single kernelized output
+                                    # series causally convoluted with respect to all other series/input window length
+                                    result[b, h, xx, y, j, f] += (
+                                        kernel[h, y, xx, j, i] * x[b, xx, i, f]
+                                    )
+        # Result:
+        # (batch_size, number_of_heads, number_of_series, number_of_series, length_input_window, hidden_dimensionality)
+        return result
```
[the remark this answers is not in the document]
Comment (boersmamarcel) [reply fintorch/models/timeseries/causalformer/CausalConvolution.py]: yes, indeed. I'll swap around the test and the execution code. I liked this version because it is more verbose (which is good for educational purposes). However, as stated here: https://pytorch.org/docs/stable/generated/torch.einsum.html einsum of more performant. Even more so, when we install a special module of PyTorch.

> Please install opt-einsum (https://optimized-einsum.readthedocs.io/en/stable/) in order to enroll into a more performant einsum. You can install when installing torch like so: pip install torch[opt-einsum] or by itself with pip install opt-einsum.

[the remark this answers is not in the document]
Comment (boersmamarcel) [reply fintorch/models/timeseries/causalformer/CausalConvolution.py]: Yes, please. That would be helpful.

[the remark this answers is not in the document]
Comment (boersmamarcel) [reply fintorch/models/timeseries/causalformer/CausalConvolution.py]: I'll add it to the project.toml file as an dependency.

File: fintorch/models/timeseries/causalformer/MultiHeadAttention.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        self.activation = nn.Softmax(dim=-1)
+        self.dropout = nn.Dropout(0.1)
+
+    def forward(
+        self, x: torch.Tensor, mask: Optional[torch.Tensor] = None
+    ) -> torch.Tensor:
+        # Make Q, K, V
+        Q, K, V = x, x, x
+        # Q, K, V: (batch_size, number_of_series, length_input_window, hidden_dimensionality)
+
```
[the remark this answers is not in the document]
Comment (boersmamarcel) [reply fintorch/models/timeseries/causalformer/MultiHeadAttention.py:89]: The linear projections are done outside the attention mechanism because we have a "special" way to project for the Value embeddings.

File: fintorch/models/timeseries/causalformer/MultiHeadAttention.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    def forward(
+        self,
+        Q: torch.Tensor,
+        K: torch.Tensor,
+        V: torch.Tensor,
+        mask: Optional[torch.Tensor] = None,
+    ) -> torch.Tensor:
+        # Q, K, (batch_size, number_of_heads, number_of_series, hidden_dimensionality)
+        # V: (batch_size, number_of_heads, number_of_series, number_of_series, length_input_window, feature_dim)
+
```
[the remark this answers is not in the document]
Comment (boersmamarcel) [reply fintorch/models/timeseries/causalformer/MultiHeadAttention.py:89]: updated docstring

File: fintorch/models/timeseries/causalformer/MultivariateCausalAttention.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+            embedding_size=embedding_size,
+            tau=tau,
+        )
+
+        self.tensor_head_dimensionality = embedding_size // number_of_heads
+        self.number_of_heads = number_of_heads
+        self.number_of_series = number_of_series
+        self.input_window = length_input_window
+        self.feature_dimensionality = feature_dimensionality
+        self.concat_proj = nn.Linear(
```
[the remark this answers is not in the document]
Comment (boersmamarcel) [reply fintorch/models/timeseries/causalformer/MultivariateCausalAttention.py]: we floor the values, hence the // division

File: fintorch/models/timeseries/causalformer/MultiHeadAttention.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        self.number_of_heads = number_of_heads
+        self.number_of_series = number_of_series
+        self.input_window = length_input_window
+        self.tau = tau
+        self.embedding_size = embedding_size
+        self.hidden_dimensionality = embedding_size // number_of_heads
+
+        self.activation = nn.Softmax(dim=-1)
+        self.dropout = nn.Dropout(0.1)
+
```
[the remark this answers is not in the document]
Comment (boersmamarcel) [reply fintorch/models/timeseries/causalformer/MultiHeadAttention.py:79]: same remark here, we floor the division with //

File: fintorch/models/timeseries/causalformer/MultivariateCausalAttention.py [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+            in_features=embedding_size,
+            out_features=embedding_size,
+            bias=False,
+        )
+
+        # Causal convolution for V, use "raw" inputs
+        self.V_proj = CausalConvolution(
+            number_of_series=number_of_series,
+            length_input_window=length_input_window,
+            number_of_heads=number_of_heads,
```
[the remark this answers is not in the document]
Comment (boersmamarcel) [reply fintorch/models/timeseries/causalformer/MultivariateCausalAttention.py:96]: I think this correct, the docstring specifies:                 v (torch.Tensor): Value tensor of shape
                    (batch_size, number_of_heads, number_of_series, length_input_window, feature_dimensionality).

File: fintorch/datasets/diamondata.py [hunk 7]
```diff
@@ -0,0 +1,403 @@
...
+            if os.path.exists(local_path):
+                os.remove(local_path)
+            raise
+        except Exception as e:
+            print(f"An error occurred during download: {e}")
+            if os.path.exists(local_path):
+                os.remove(local_path)
+            raise
+    else:
+        print(f"Data file already exists at {local_path}. Skipping download.")
```
[the remark this answers is not in the document]
Comment (boersmamarcel) [reply fintorch/datasets/diamondata.py]: #115

File: examples/datasets/diamonds.py [hunk 8]
```diff
@@ -0,0 +1,101 @@
+from fintorch.datasets.diamondata import DiamondDataModule
+
+if __name__ == "__main__":
+    # Define parameters
+    TIME_STEP = 16
+    OUTPUT_WINDOW = 15
+    BATCH_SIZE = 32
+    # Use the current directory to find the data file
+    DATA_PATH = "data_0.csv"
+    DUMMY_DATA_ROWS = 200  # Number of rows for dummy data
```
[the remark this answers is not in the document]
Comment (boersmamarcel) [reply examples/datasets/diamonds.py:10]: #115

File: tests/models/causalformer/test_encoder.py [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+
+    # Forward pass
+    x_emb = encoder.embedding(x)
+
+    # Assertions
+    expected_shape = (batch_size, number_of_series, embedding_size)
+    assert (
+        x_emb.shape == expected_shape
+    ), f"Expected embedding output shape {expected_shape}, but got {x_emb.shape}"
+    assert isinstance(x_emb, torch.Tensor), "Embedding output should be a torch.Tensor"
```
[the remark this answers is not in the document]
Comment (boersmamarcel) [reply tests/models/causalformer/test_encoder.py:95]: not important to optimize the test code, skipping it

File: tests/models/causalformer/test_encoder.py [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+            length_input_window,
+            feature_dimensionality,
+        )
+        assert (
+            output.shape == expected_shape
+        ), f"Expected output shape {expected_shape}, but got {output.shape}"
+        assert isinstance(output, torch.Tensor), "Output should be a torch.Tensor"
+
+    except Exception as e:
+        raise AssertionError(f"Encoder forward pass failed with exception: {e}")
```
[the remark this answers is not in the document]
Comment (boersmamarcel) [reply tests/models/causalformer/test_encoder.py:53]: We have a new input types tensor, see time-series.md. It is not perfectly implemented yet everywhere, but we are getting there.
